File: sdk/java/src/main/java/io/kcm/examples/BasicCrud.java
package io.kcm.examples;

import io.kcm.*;

/**
 * KCM Java SDK — Basic CRUD Example.
 *
 * Demonstrates: insert, query, update, delete operations on facts.
 */
public class BasicCrud {
    public static void main(String[] args) throws KcmException {
        System.out.println("=== KCM Java SDK — Basic CRUD Example ===\n");

        try (KcmDatabase db = new KcmDatabase()) {

            // --- INSERT ---
            System.out.println("--- Insert Facts ---");
            db.insert(new Fact(1, (byte) 0, 2, 0.95,
                    (byte) 0, 0L, (byte) 0, 1, (byte) 0, (short) 0));
            db.insert(new Fact(2, (byte) 1, 3, 0.90,
                    (byte) 1, 0L, (byte) 1, 1, (byte) 0, (short) 1));
            db.insert(new Fact(3, (byte) 2, 4, 0.85,
                    (byte) 2, 0L, (byte) 2, 1, (byte) 0, (short) 2));
            db.insert(new Fact(1, (byte) 3, 5, 0.80,
                    (byte) 3, 0L, (byte) 2, 1, (byte) -1, (short) 7));
            System.out.printf("  Inserted 4 facts, Total: %d, Active: %d%n",
                    db.factCount(), db.activeFactCount());

            // --- QUERY ALL ---
            System.out.println("\n--- Query All Facts ---");
            try (KcmQuery query = db.query("SELECT * FROM facts")) {
                while (query.hasNext()) {
                    Fact f = query.next();
                    System.out.printf("  Subject: %d, Predicate: %d, Object: %d, Confidence: %.2f%n",
                            f.subject, f.predicate, f.object, f.confidence);
                }
            }

            // --- QUERY WITH FILTER ---
            System.out.println("\n--- KQL Query: SELECT * FROM facts WHERE subject = 1 ---");
            try (KcmQuery query = db.query("SELECT * FROM facts")) {
                int count = 0;
                while (query.hasNext()) {
                    Fact f = query.next();
                    if (f.subject == 1) {
                        System.out.printf("  Subject: %d, Predicate: %d, Object: %d%n",
                                f.subject, f.predicate, f.object);
                        count++;
                    }
                }
                System.out.printf("  Found %d facts with subject=1%n", count);
            }

            // --- UPDATE ---
            System.out.println("\n--- Update Fact ---");
            db.update(0, new Fact(10, (byte) 0, 20, 0.99,
                    (byte) 5, 0L, (byte) 3, 2, (byte) 2, (short) 10));
            System.out.println("  Updated row 0: subject changed to 10");

            // --- DELETE ---
            System.out.println("\n--- Delete Fact ---");
            db.delete(3);
            System.out.printf("  Deleted row 3, Total: %d, Active: %d%n",
                    db.factCount(), db.activeFactCount());

            // --- VERIFY COUNTS ---
            System.out.println("\n--- Verify Counts ---");
            assert db.factCount() == 4 : "Expected 4 total, got " + db.factCount();
            assert db.activeFactCount() == 3 : "Expected 3 active, got " + db.activeFactCount();
            System.out.println("  Counts verified: 4 total, 3 active");
        }

        System.out.println("\n=== All operations completed ===");
    }
}

File: sdk/dotnet/examples/standalone/BasicCrud.cs
using System;
using Kcm;

/// <summary>
/// KCM .NET SDK — Basic CRUD Example.
/// Demonstrates: insert, query, update, delete operations on facts.
/// </summary>
namespace Kcm.Examples
{
    public static class BasicCrud
    {
        public static void Run()
        {
            Console.WriteLine("=== KCM .NET SDK — Basic CRUD Example ===\n");

            using var db = new KcmDatabase();

            // --- INSERT ---
            Console.WriteLine("--- Insert Facts ---");
            db.Insert(new Fact(1, 0, 2, 0.95, evidence: 0, timestamp: 0, context: 0, version: 1, priority: 0, owner: 0));
            db.Insert(new Fact(2, 1, 3, 0.90, evidence: 1, timestamp: 0, context: 1, version: 1, priority: 0, owner: 1));
            db.Insert(new Fact(3, 2, 4, 0.85, evidence: 2, timestamp: 0, context: 2, version: 1, priority: 0, owner: 2));
            db.Insert(new Fact(1, 3, 5, 0.80, evidence: 3, timestamp: 0, context: 2, version: 1, priority: -1, owner: 7));
            Console.WriteLine($"  Inserted 4 facts, Total: {db.FactCount()}, Active: {db.ActiveFactCount()}");

            // --- QUERY ALL ---
            Console.WriteLine("\n--- Query All Facts ---");
            using (var query = db.Query("SELECT * FROM facts"))
            {
                foreach (var f in query)
                {
                    Console.WriteLine($"  Subject: {f.Subject}, Predicate: {f.Predicate}, Object: {f.Object}, Confidence: {f.Confidence:F2}");
                }
            }

            // --- QUERY WITH FILTER ---
            Console.WriteLine("\n--- KQL Query: SELECT * FROM facts WHERE subject = 1 ---");
            using (var query = db.Query("SELECT * FROM facts"))
            {
                int count = 0;
                foreach (var f in query)
                {
                    if (f.Subject == 1)
                    {
                        Console.WriteLine($"  Subject: {f.Subject}, Predicate: {f.Predicate}, Object: {f.Object}");
                        count++;
                    }
                }
                Console.WriteLine($"  Found {count} facts with subject=1");
            }

            // --- UPDATE ---
            Console.WriteLine("\n--- Update Fact ---");
            db.Update(0, new Fact(10, 0, 20, 0.99, evidence: 5, timestamp: 0, context: 3, version: 2, priority: 2, owner: 10));
            Console.WriteLine("  Updated row 0: subject changed to 10");

            // --- DELETE ---
            Console.WriteLine("\n--- Delete Fact ---");
            db.Delete(3);
            Console.WriteLine($"  Deleted row 3, Total: {db.FactCount()}, Active: {db.ActiveFactCount()}");

            // --- VERIFY COUNTS ---
            Console.WriteLine("\n--- Verify Counts ---");
            Console.WriteLine($"  Total: {db.FactCount()}, Active: {db.ActiveFactCount()}");

            Console.WriteLine("\n=== All operations completed ===");
        }
    }
}
